pkg/daemon/daemon.go

```diff
@@ -177,38 +234,123 @@ func (dn *Daemon) CheckStateOnBoot(stop <-chan struct{}) error {
+		if err != nil {
+			return err
+		}
+		machineConfig, err = dn.getMachineConfigFromFile(absoluteOnceFrom)
+		if err != nil {
+			return err
+		}
+		// Execute update without hitting the cluster
+		return dn.update(&oldConfig, machineConfig)
+	}
+	// Otherwise return an error as the input format is unsupported
+	return fmt.Errorf("%s is not a path nor url; can not run once", dn.onceFrom)
+}
+
+// handleNodeUpdate is the gatekeeper handler for informer callbacks detecting
+// node changes. If an update is requested by the controller, we assume that
+// that means something changed and pass over to execution methods no matter what.
 // Also note that we only care about node updates, not creation or deletion.
 func (dn *Daemon) handleNodeUpdate(old, cur interface{}) {
 	node := cur.(*corev1.Node)
 
 	// First check if the node that was updated is this daemon's node
-	if (node.Name != dn.name) {
-		// The node that was changed was not ours
-		return
+	if node.Name == dn.name {
+		// Pass to the shared update prep method
+		needUpdate, err := dn.prepUpdateFromCluster()
+		if err != nil {
+			// On prepUpdateFromCluster error the node should already be marked degraded
+			glog.V(2).Infof("Unable to prep update: %s", err)
+			dn.reboot()
+			return
+		}
+		// Only executeUpdateFromCluster when we need to update
+		if needUpdate {
+			// Note that if executeUpdateFromCluster errors it will mark the node
+			// degraded and reboot.
+			if err = dn.executeUpdateFromCluster(); err != nil {
+				glog.V(2).Infof("Unable to execute update: %s", err)
+				return
+			}
+		}
 	}
+	// The node that was changed was not ours, return out
+	return
+}
 
+// prepUpdateFromCluster handles the shared update prepping functionality for
+// flows that expect the cluster to already be available.
+func (dn *Daemon) prepUpdateFromCluster() (bool, error) {
 	// Then check we're not already in a degraded state.
 	if state, err := getNodeAnnotation(dn.kubeClient.CoreV1().Nodes(), dn.name, MachineConfigDaemonStateAnnotationKey); err != nil {
 		// try to set to degraded... because we failed to check if we're degraded
 		glog.Errorf("Marking degraded due to: %v", err)
 		setUpdateDegraded(dn.kubeClient.CoreV1().Nodes(), dn.name)
-		return
+		return false, err
 	} else if state == MachineConfigDaemonStateDegraded {
 		// Just return since we want to continue sleeping
-		return
+		return false, fmt.Errorf("state is already degraded")
+	}
+
+	// Grab the node instance
+	node, err := dn.kubeClient.CoreV1().Nodes().Get(dn.name, metav1.GetOptions{})
+	if err != nil {
+		return false, err
 	}
 
 	// Detect if there is an update
-	if (node.Annotations[DesiredMachineConfigAnnotationKey] == node.Annotations[CurrentMachineConfigAnnotationKey]) {
+	if node.Annotations[DesiredMachineConfigAnnotationKey] == node.Annotations[CurrentMachineConfigAnnotationKey] {
 		// No actual update to the config
-		return
+		glog.V(2).Info("No updating is required")
```

Comment:
I suggest removing this since this will likely get logged a lot in cluster operation.